pkg/resource/v1/certconfig/desired.go

```diff
@@ -2,8 +2,358 @@ package certconfig
+				ClusterID:           clusterConfig.ClusterID,
+				CommonName:          clusterConfig.Domain.Prometheus,
+				DisableRegeneration: false,
+				TTL:                 clusterConfig.CertTTL,
+			},
+			VersionBundle: v1alpha1.CertConfigSpecVersionBundle{
+				Version: clusterConfig.VersionBundleVersion,
+			},
+		},
+	}
+}
+
+func newServiceAccountCertConfig(clusterConfig *cluster.Config, cert certs.Cert, projectName string) *v1alpha1.CertConfig {
+	certName := string(cert)
+	return &v1alpha1.CertConfig{
+		TypeMeta: apimetav1.TypeMeta{
+			Kind:       certKind,
+			APIVersion: certAPIVersion,
+		},
+		ObjectMeta: apimetav1.ObjectMeta{
+			Name: key.CertConfigName(clusterConfig.ClusterID, cert),
+			Labels: map[string]string{
+				label.Cluster:         clusterConfig.ClusterID,
+				label.LegacyClusterID: clusterConfig.ClusterID,
+				label.LegacyComponent: certName,
+				label.ManagedBy:       projectName,
+				label.Organization:    clusterConfig.Organization,
+			},
+		},
+		Spec: v1alpha1.CertConfigSpec{
+			Cert: v1alpha1.CertConfigSpecCert{
+				AllowBareDomains:    false,
+				ClusterComponent:    certName,
+				ClusterID:           clusterConfig.ClusterID,
+				CommonName:          clusterConfig.Domain.ServiceAccount,
+				DisableRegeneration: false,
+				TTL:                 clusterConfig.CertTTL,
+			},
+			VersionBundle: v1alpha1.CertConfigSpecVersionBundle{
+				Version: clusterConfig.VersionBundleVersion,
+			},
+		},
+	}
+}
+
+func newWorkerCertConfig(clusterConfig *cluster.Config, cert certs.Cert, projectName string) *v1alpha1.CertConfig {
+	certName := string(cert)
+	return &v1alpha1.CertConfig{
+		TypeMeta: apimetav1.TypeMeta{
+			Kind:       certKind,
+			APIVersion: certAPIVersion,
+		},
+		ObjectMeta: apimetav1.ObjectMeta{
+			Name: key.CertConfigName(clusterConfig.ClusterID, cert),
+			Labels: map[string]string{
+				label.Cluster:         clusterConfig.ClusterID,
+				label.LegacyClusterID: clusterConfig.ClusterID,
+				label.LegacyComponent: certName,
+				label.ManagedBy:       projectName,
+				label.Organization:    clusterConfig.Organization,
+			},
+		},
+		Spec: v1alpha1.CertConfigSpec{
+			Cert: v1alpha1.CertConfigSpecCert{
+				AllowBareDomains:    true,
+				AltNames:            kubeAltNames,
+				ClusterComponent:    certName,
+				ClusterID:           clusterConfig.ClusterID,
+				CommonName:          clusterConfig.Domain.Worker,
+				DisableRegeneration: false,
+				TTL:                 clusterConfig.CertTTL,
+			},
+			VersionBundle: v1alpha1.CertConfigSpecVersionBundle{
+				Version: clusterConfig.VersionBundleVersion,
+			},
+		},
+	}
+}
+
+func newServerDomain(apiEndpoint string, cert certs.Cert) (string, error) {
+	u, err := url.Parse(apiEndpoint)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	splitted := strings.Split(u.Host, ".")
+
+	// This is the subdomain part.
+	splitted[0] = string(cert)
+	serverDomain := strings.Join(splitted, ".")
+
+	return serverDomain, nil
 }
```

Comment:
I think this is something we should change in the future and replace that with some kind of base domain. Prefixes doesn't differ between clusters as it shows and this gives false feeling we can configure prefix for K8s api.

Reply:
Would it actually be feasible to already implement this as you suggest by replacing Kubernetes API Endpoint with referred commonDomain and build from there. I think all relevant logic belongs to cluster-operator and therefore it would be already doable. Or is this exactly what you meant?

Reply:
I would vote to concentrate on getting something done and refactor in smaller steps later on.

Reply:
Actually it would be good to do it now. It would save effort on migration later.

Reply:
Disagree and commit.

Reply:
Giving it a second thought that would also mean we'd have to change {aws,azure,kvm}-operator in one go. So it won't save any migration effort. I thought it only affect cluster-operator but it isn't true. So I'm with @xh3b4sd. Better to do it slowly.

Reply:
Ok. To make sure: This should be left as-is for now and refactored later together with {aws,azure,kvm}-operators?

Reply:
I think you can change ClusterConfig now. Why to postpone it? It will be created to cluster-service which is subject to change anyway.

Reply:
So basically changing [this](https://github.com/giantswarm/apiextensions/blob/master/pkg/apis/core/v1alpha1/cluster_guest_types.go#L4) `ClusterGuestConfig.API.Endpoint` to be `ClusterGuestConfig.CommonDomain` and reflect changes here?

Reply:
That sounds like a plan :)

Reply:
CRD updated, sub-domain construction simplified + added tests.